woocommerce/woocommerce
# Support additional fields in the Contact step for Checkout

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

This PR adds support for additional fields in the Contact step for Checkout. For now, fields can only be saved to the order not the customer, in a follow up issue, we will discuss the merits of that.

Closes https://github.com/woocommerce/woocommerce/issues/42082



### How to test the changes in this Pull Request:

1. Make sure you're running an experimental build, via checking the `plugins/woocommerce/blocks.ini` file and seeing if it's:
```
woocommerce_blocks_phase = 3
woocommerce_blocks_env = development
```
2. In a snippet plugin, add the following code:
```php
add_action(
	'woocommerce_loaded',
	function() {
		woocommerce_blocks_register_checkout_field(
			array(
				'id'       => 'namespace/gov-id',
				'label'    => 'Government ID',
				'location' => 'contact',
				'required' => false,
			)
		);

		woocommerce_blocks_register_checkout_field(
			array(
				'id'       => 'namespace/type-of-contact',
				'label'    => 'Type of contact',
				'location' => 'contact',
				'type'     => 'select',
				'options'  => array(
					array(
						'label' => 'Company',
						'value' => 'company',
					),
					array(
						'label' => 'Individual',
						'value' => 'individual',
					),
				),
			)
		);

		woocommerce_blocks_register_checkout_field(
			array(
				'id'       => 'namespace/marketing-opt-in',
				'label'    => 'Do you want to subscribe to our newsletter?',
				'location' => 'contact',
				'type'     => 'checkbox',
			)
		);
	}
);
```
3. In Checkout, fill out the select only, but don't touch the rest.
4. Place your order, it should pass fine.
5. Go back to Checkout, fill all the fields and check the checkbox.
6. Place the order, it should pass fine.
7. In the code snippet, change the value of `location` to `address` in all the fields.
8. Do the same steps again in Checkout, ensure no errors surface.



### Screenshot
<img width="664" alt="image" src="https://github.com/woocommerce/woocommerce/assets/6165348/8fdcb01c-03fd-4817-a385-5b5eebe6c2b2">

### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
